### ISSUE TITLE
Host listen port inappropriately held open after docker restart (with 3-part port option)

### ISSUE THREAD

**Issue by rfay** [docker/for-mac#5649]
- [x] I have tried with the latest version of Docker Desktop
  - [x] I have tried disabling enabled experimental features
  - [x] I have uploaded Diagnostics
  - Diagnostics ID: 165EF212-DB99-4FE1-90A4-E516A309B23F/20210503182644 (taken after restart with port held open)

### Expected behavior

A Docker Desktop restart automatically stops all containers and should also disconnect any port listens. 

### Actual behavior

Instead, with Docker Desktop 3.3.1 (possibly earlier versions as well), a port will still be unavailable (occupied) after a docker desktop restart.

### Information

This is easily reproducible on both arm64 and amd64 as below. It does not seem to be a problem on Windows (regular or WSL2, and it does not seem to be a problem on Linux.

  - macOS Version:  11.3
  - Intel chip or Apple chip: Same behavior with both M1 and Intel, amd64 and arm64
  - Docker Desktop Version: 3.3.1

### Steps to reproduce the behavior

1. `docker run --name nginx -d -p 127.0.0.1:80:80 nginx`
2. `curl -I http://localhost` or `nc -vz localhost 80` will show that docker is listening on port 80.
3. Restart docker desktop using the menu item
4. `nc -vz localhost 80` will show that the listen port is still occupied. This means that the container cannot be restarted, nor can another container occupy port 80. As a result, docker desktop has to be restarted *again* to get port 80 cleared.

Removing the container with `docker rm nginx` does not solve this. You still have to restart docker desktop.

This doesn't seem to happen when binding to 0.0.0.0, so `docker run --name nginx -d -p 80:80 nginx` doesn't seem to cause this behavior.

**Comment by rfay** [comment]
This looks related to https://github.com/docker/for-mac/issues/5611 - I'll test 3.3.2 to see if it's better.

**Comment by rfay** [comment]
This is not fixed in 3.3.2, would love it if you could take a look @djs55 - I'm getting support issues from so many users.

**Comment by djs55** [comment]
@rfay thanks for the ping, I'll take a look.

**Comment by djs55** [comment]
I think I understand the issue. Go takes care to set the "close on exec" flag on new file descriptors to ensure they aren't accidentally inherited by new subprocesses, which is good. Meanwhile modern macOS allows a regular process to bind a privileged port (like `80`) to any (`0.0.0.0`) IP. However in the case of a privileged port bound to a specific IP (like `127.0.0.1`) we still need to use a privileged helper process to bind the port, and then send the fd to the `com.docker.backend` process by `sendmsg`. Unfortunately the `recvmsg` doesn't have the "close on exec" flag set. The "Restart" option then asks the `com.docker.backend` process to restart `hyperkit`/`qemu` etc and they end up inheriting the listening file descriptor. When the `com.docker.backend` closes the port it doesn't really close because the child processes still have it open.

**Comment by djs55** [comment]
@rfay I've got a build with a candidate fix if you'd like to try it: https://desktop-stage.docker.com/mac/stable/amd64/64153/Docker.dmg (replace `amd64` with `arm64` if you'd prefer the Apple Silicon version)

I used your example `docker run --name nginx -d -p 127.0.0.1:80:80 nginx` and listed the port with:
```
% lsof -sTCP:LISTEN | grep http | grep LISTEN
com.docke 39415 dave  105u     IPv6 0xb79ae126db709abd         0t0                 TCP *:http (LISTEN)
```
After clicking on Restart and waiting, the port disappeared (as expected). Previously it was held open by lots of subprocesses.

**Comment by rfay** [comment]
Build 64153 on arm64 worked perfectly for me, solved the problem described in the OP here and in ddev at https://github.com/drud/ddev/issues/2981 

I've installed it on a test runner and will run a full test run with it, but I don't expect any issues. 

THANK YOU!

**Comment by rfay** [comment]
@djs55 I don't see this mentioned in the [3.4.0 release notes](https://docs.docker.com/docker-for-mac/release-notes/#docker-desktop-340) Did it make it in?

**Comment by mat007** [comment]
@rfay yes it did, it’s
> Avoid leaking open ports bound to privileged ports and specific IPs over engine Restart. Fixes docker/for-mac#5649.
